### PR TITLE
Add `rows_*()` verbs

### LINKED ISSUE #736: Add `rows_*.tbl_dbi()` methods
These methods currently live in the `dm` package but it would make more sense to have them in `dbplyr`.

### PR DESCRIPTION
**Pull request by mgirlich**
Closes #736.

`returning`: This argument is still experimental and not (yet?) in the generic. ~To support `tidyselect` and not trigger a warning by `rlang::check_dots_used()` (in the generic) the user has to quote it manually (see https://github.com/tidyverse/dplyr/issues/5985).~ (Edit: not necessary anymore after `check_dots_used()` was removed in the generic).
Theoretically, one could execute arbitrary SQL in the `RETURNING` clause but I'm not sure whether there are sufficient use cases to support this.

`nonexistent`: The current implementation is `nonexistent = "ignore"`. To support `nonexistent = "error"` one could use a foreign key constraint on `y`. I'll tackle this in another PR. (Update: renamed to `unmatched` and "error" is not supported for now)

`rows_upsert()`: This is the most tricky and needs different methods:
* Some databases (e.g. Oracle or MS SQL) support `MERGE INTO ... WHEN (NOT) MATCHED`
* Some databases (e.g. SQLite, Postgres) have an upsert if there is a unique constraint, basically `INSERT ... ON CONFLICT DO UPDATE/NOTHING`
* If there is no unique constraint this can still be done in one query with a CTE (e.g. Postgres) or two queries (e.g. SQLite)

It might make sense to add an argument so that the user can choose whether a unique constraint exists or not. See
https://github.com/cynkra/dm/pull/616#issuecomment-920613435 for some more information on the Pros/Cons. I will do this in another PR.

### CODE REVIEW

File: R/rows.R
Comment (mgirlich) [inline R/rows.R:41]: I think it should be possible to mute this message. But how? Add another argument `quiet`? And should this inform how to add an index via `db_create_index()`?

Comment (mgirlich) [reply R/rows.R:41]: It would also be possible to add an option `check` that collects the data and then checks this but I'm not really sure this is worth it as the unique index makes much more sense.

**Comment by mgirlich** [comment]
@hadley It would be great to merge most of the remaining PRs and then make a CRAN release. Do you think we can manage that in the next couple of weeks?

**Review by hadley — COMMENTED** [review]
Is there anything in particular you'd like me to look at? I'm unlikely to have time to review it depth, but I trust that you've implemented it correctly (and we'll discover and fix any mistakes as people start to use it).

**Comment by krlmlr** [comment]
I'll take a closer look here and double-check if it's consistent/compatible with what's in dm. It might be worth to release a dm version that uses the dbplyr implementation if it's available before releasing dbplyr.

File: R/rows.R
Comment (krlmlr) [reply R/rows.R:41]: `conflict = "allow"` or `conflict = "silent"` ?

File: R/rows.R
Comment (krlmlr) [inline R/rows.R:491]: I don't think we should be doing anything here. We have tools in dm to check referential integrity and key constraints "after the fact", and we can simulate with `in_place = FALSE` . Actively checking here can be very expensive.

File: _pkgdown.yml
Comment (krlmlr) [inline _pkgdown.yml:35]: There is one help page for all "rows" functions, right?

```suggestion
  - "rows-db"
```

File: R/db-sql.R
Comment (krlmlr) [inline R/db-sql.R:142]: Looks like this was accidentally moved up to `sql_returning_cols()`:

```suggestion
#' @export
sql_random <- function(con) {
```

**Review by krlmlr — COMMENTED** [review]
Thanks, looks great -- using the internal dbplyr tools surely gives much better code than using `paste()` in dm. This is the first review round, looking only at code. Now installing the branch and checking compatibility with dm.

File: R/rows.R
Comment (krlmlr) [reply R/rows.R:41]: I've proposed to change this to `conflict = "drop"` in the dplyr PR, and have `conflict = "ignore"` allow duplicate row. Hope we can reach a consensus.

File: R/rows.R
Comment (krlmlr) [inline R/rows.R:58]: We still could run a query like `union_all() %>% count() %>% filter(n > 1) %>% head(1) %>% collect()` and error if we see at least one row here, along with a message that this is expensive and users can use `conflict = "ignore"` (meaning that conflicts are allowed). The alternative `conflict = "drop"` can also be expensive on the database.

**Comment by krlmlr** [comment]
Are you planning to support the `conflict`/`nonexistent` arguments for all verbs? How about the following simplified approach:

- We include the argument with the same defaults as in dplyr
- Any other value than the "natural" thing on the database (i.e. `conflict = "ignore"` meaning "allow duplicate rows", or `nonexistent = "ignore"`) gives a warning once every 8 hours (via `rlang::warn(.frequency = "regularly")`)
- Exception: `rows_insert(conflict = "drop")` already has code, we can keep using that

This behavior is consistent with `sum(na.rm = TRUE)` which also doesn't work on databases. I also don't mind an error instead of a warning, as long as there's a way to support what's natural on the database. Later we could work on perhaps supporting other options for `conflict` and `nonexistent` .

Of course this means that dplyr must support what's natural on the database.

**Comment by krlmlr** [comment]
Otherwise, compatibility with dm looks great: https://github.com/cynkra/dm/pull/912 succeeds locally, now testing on databases.

File: R/lazy-ops.R
Comment (mgirlich) [reply R/lazy-ops.R:34]: Hmm, I was pretty sure I already added `...` in `lazy_query_base()` at some point 🤷 But yeah, makes sense

File: R/db-sql.R
Comment (mgirlich) [reply R/db-sql.R:497]: I think it should work in many databases but I also think most offer better or easier proprietary alternatives (`MERGE`, `INSERT INTO ... ON CONFLICT`). Even in 2022 upserts are a bit of a mess in SQL...

File: R/rows.R
Comment (mgirlich) [reply R/rows.R:58]: I think it is worth supporting `conflict = "error"` but `collecting()` by default does not sound good to me. So, maybe `conflict` should not be an optional argument for `rows_insert.tbl_lazy(in_place = FALSE)`.

File: R/rows.R
Comment (mgirlich) [reply R/rows.R:491]: What do you mean by "after the fact"?

There should definitely be tools to easiliy check the uniqueness of `y` and they should be well-documented.

If `y` is a local table the check is easy, so it would be an option to check this before copying.

File: R/rows.R
Comment (krlmlr) [reply R/rows.R:58]: We pay extra for the checks in dplyr too, and we collect at most one row. We could give a message if the `conflict` argument is missing, to ensure that users are aware?

**Comment by krlmlr** [comment]
Unfortunately, double quoting is an issue with dm. We use `ident_q` to mark identifiers as quoted, this requires using `escape()` instead of `sql_escape_ident()` -- also in existing code. Figuring out the best way to handle this.

**Comment by krlmlr** [comment]
Do we want a message if `in_place` is left unspecified?

**Comment by krlmlr** [comment]
There are several ways to create a writable `tbl` object to a remote table: by name, via `in_schema()`, `ident()`, or `ident_q()` . All these work with `rows_insert()` and SQLite, not so much with SQL Server (and Postgres, judging by the test failures in dm). I'll take a closer look.

## SQLite

``` r
library(dbplyr)
library(tidyverse)

con <- src_memdb()$con

con_frame <- function(...) {
  copy_inline(con, tibble(...))
}

DBI::dbExecute(con, "CREATE TABLE test (a INTEGER)")
#> [1] 0

tbl_pure <- tbl(con, "test")
tbl_pure
#> # Source:   table<test> [?? x 1]
#> # Database: sqlite 3.38.0 [:memory:]
#> # … with 1 variable: a <int>
rows_insert(tbl_pure, memdb_frame(a = 1L), conflict = "ignore")
#> Matching, by = "a"
#> # Source:   [?? x 1]
#> # Database: sqlite 3.38.0 [:memory:]
#>       a
#>   <int>
#> 1     1
rows_insert(tbl_pure, memdb_frame(a = 1L), conflict = "ignore", in_place = TRUE)
#> Matching, by = "a"

tbl_in_schema <- tbl(con, in_schema("main", "test"))
tbl_in_schema
#> # Source:   table<`main`.`test`> [?? x 1]
#> # Database: sqlite 3.38.0 [:memory:]
#>       a
#>   <int>
#> 1     1
rows_insert(tbl_in_schema, memdb_frame(a = 2L), conflict = "ignore")
#> Matching, by = "a"
#> # Source:   [?? x 1]
#> # Database: sqlite 3.38.0 [:memory:]
#>       a
#>   <int>
#> 1     1
#> 2     2
rows_insert(tbl_in_schema, memdb_frame(a = 2L), conflict = "ignore", in_place = TRUE)
#> Matching, by = "a"

tbl_ident <- tbl(con, ident("test"))
tbl_ident
#> # Source:   table<test> [?? x 1]
#> # Database: sqlite 3.38.0 [:memory:]
#>       a
#>   <int>
#> 1     1
#> 2     2
rows_insert(tbl_ident, memdb_frame(a = 3L), conflict = "ignore")
#> Matching, by = "a"
#> # Source:   [?? x 1]
#> # Database: sqlite 3.38.0 [:memory:]
#>       a
#>   <int>
#> 1     1
#> 2     2
#> 3     3
rows_insert(tbl_ident, memdb_frame(a = 3L), conflict = "ignore", in_place = TRUE)
#> Matching, by = "a"

tbl_ident_q <- tbl(con, ident_q("`test`"))
tbl_ident_q
#> # Source:   table<`test`> [?? x 1]
#> # Database: sqlite 3.38.0 [:memory:]
#>       a
#>   <int>
#> 1     1
#> 2     2
#> 3     3
rows_insert(tbl_ident_q, memdb_frame(a = 4L), conflict = "ignore")
#> Matching, by = "a"
#> # Source:   [?? x 1]
#> # Database: sqlite 3.38.0 [:memory:]
#>       a
#>   <int>
#> 1     1
#> 2     2
#> 3     3
#> 4     4
rows_insert(tbl_ident_q, memdb_frame(a = 4L), conflict = "ignore", in_place = TRUE)
#> Matching, by = "a"
```

<sup>Created on 2022-04-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

## SQL Server (with #823)

``` r
library(dbplyr)
library(tidyverse)

# FIXME: Specify server connection:
con <- sql_server_con()

con_frame <- function(...) {
  copy_inline(con, tibble(...))
}
con_frame(a = 1)
#> # Source:   [?? x 1]
#> # Database: Microsoft SQL Server 15.00.4198
#>       a
#>   <dbl>
#> 1     1

DBI::dbExecute(con, "CREATE TABLE test (a INTEGER)")
DBI::dbExecute(con, "DELETE FROM test")
#> [1] 0

tbl_pure <- tbl(con, "test")
tbl_pure
#> # Source:   table<test> [?? x 1]
#> # Database: Microsoft SQL Server 15.00.4198
#> # … with 1 variable: a <int>
rows_insert(tbl_pure, con_frame(a = 1L), conflict = "ignore")
#> Matching, by = "a"
#> # Source:   [?? x 1]
#> # Database: Microsoft SQL Server 15.00.4198[@SV0376/test]
#>       a
#>   <int>
#> 1     1
rows_insert(tbl_pure, con_frame(a = 1L), conflict = "ignore", in_place = TRUE)
#> Matching, by = "a"

tbl_in_schema <- tbl(con, in_schema("dbo", "test"))
tbl_in_schema
#> # Source:   table<"dbo"."test"> [?? x 1]
#> # Database: Microsoft SQL Server 15.00.4198
#>       a
#>   <int>
#> 1     1
rows_insert(tbl_in_schema, con_frame(a = 2L), conflict = "ignore")
#> Matching, by = "a"
#> # Source:   [?? x 1]
#> # Database: Microsoft SQL Server 15.00.4198
#>       a
#>   <int>
#> 1     1
#> 2     2
rows_insert(tbl_in_schema, con_frame(a = 2L), conflict = "ignore", in_place = TRUE)
#> Matching, by = "a"
#> Error in new_result(connection@ptr, statement, immediate): nanodbc/nanodbc.cpp:1594: 00000: [FreeTDS][SQL Server]The multi-part identifier ""dbo"."test".a" could not be bound.

tbl_ident <- tbl(con, ident("test"))
tbl_ident
#> # Source:   table<test> [?? x 1]
#> # Database: Microsoft SQL Server 15.00.4198
#>       a
#>   <int>
#> 1     1
rows_insert(tbl_ident, con_frame(a = 3L), conflict = "ignore")
#> Matching, by = "a"
#> # Source:   [?? x 1]
#> # Database: Microsoft SQL Server 15.00.4198
#>       a
#>   <int>
#> 1     1
#> 2     3
rows_insert(tbl_ident, con_frame(a = 3L), conflict = "ignore", in_place = TRUE)
#> Matching, by = "a"

tbl_ident_q <- tbl(con, ident_q("[test]"))
tbl_ident_q
#> # Source:   table<[test]> [?? x 1]
#> # Database: Microsoft SQL Server 15.00.4198
#>       a
#>   <int>
#> 1     1
#> 2     3
rows_insert(tbl_ident_q, con_frame(a = 4L), conflict = "ignore")
#> Matching, by = "a"
#> # Source:   [?? x 1]
#> # Database: Microsoft SQL Server 15.00.4198
#>       a
#>   <int>
#> 1     1
#> 2     3
#> 3     4
rows_insert(tbl_ident_q, con_frame(a = 4L), conflict = "ignore", in_place = TRUE)
#> Matching, by = "a"
#> Error in new_result(connection@ptr, statement, immediate): nanodbc/nanodbc.cpp:1594: 00000: [FreeTDS][SQL Server]The multi-part identifier "[test].a" could not be bound.
```

<sup>Created on 2022-04-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by krlmlr** [comment]
There are several ways to create a writable `tbl` object to a remote table: by name, via `in_schema()`, `ident()`, or `ident_q()` . All these work with `rows_insert()` and SQLite, not so much with SQL Server (and Postgres, judging by the test failures in dm). I'll take a closer look.

## SQLite

``` r
library(dbplyr)
library(tidyverse)

con <- src_memdb()$con

con_frame <- function(...) {
  copy_inline(con, tibble(...))
}

DBI::dbExecute(con, "CREATE TABLE test (a INTEGER)")
#> [1] 0

tbl_pure <- tbl(con, "test")
tbl_pure
#> # Source:   table<test> [?? x 1]
#> # Database: sqlite 3.38.0 [:memory:]
#> # … with 1 variable: a <int>
rows_insert(tbl_pure, memdb_frame(a = 1L), conflict = "ignore")
#> Matching, by = "a"
#> # Source:   [?? x 1]
#> # Database: sqlite 3.38.0 [:memory:]
#>       a
#>   <int>
#> 1     1
rows_insert(tbl_pure, memdb_frame(a = 1L), conflict = "ignore", in_place = TRUE)
#> Matching, by = "a"

tbl_in_schema <- tbl(con, in_schema("main", "test"))
tbl_in_schema
#> # Source:   table<`main`.`test`> [?? x 1]
#> # Database: sqlite 3.38.0 [:memory:]
#>       a
#>   <int>
#> 1     1
rows_insert(tbl_in_schema, memdb_frame(a = 2L), conflict = "ignore")
#> Matching, by = "a"
#> # Source:   [?? x 1]
#> # Database: sqlite 3.38.0 [:memory:]
#>       a
#>   <int>
#> 1     1
#> 2     2
rows_insert(tbl_in_schema, memdb_frame(a = 2L), conflict = "ignore", in_place = TRUE)
#> Matching, by = "a"

tbl_ident <- tbl(con, ident("test"))
tbl_ident
#> # Source:   table<test> [?? x 1]
#> # Database: sqlite 3.38.0 [:memory:]
#>       a
#>   <int>
#> 1     1
#> 2     2
rows_insert(tbl_ident, memdb_frame(a = 3L), conflict = "ignore")
#> Matching, by = "a"
#> # Source:   [?? x 1]
#> # Database: sqlite 3.38.0 [:memory:]
#>       a
#>   <int>
#> 1     1
#> 2     2
#> 3     3
rows_insert(tbl_ident, memdb_frame(a = 3L), conflict = "ignore", in_place = TRUE)
#> Matching, by = "a"

tbl_ident_q <- tbl(con, ident_q("`test`"))
tbl_ident_q
#> # Source:   table<`test`> [?? x 1]
#> # Database: sqlite 3.38.0 [:memory:]
#>       a
#>   <int>
#> 1     1
#> 2     2
#> 3     3
rows_insert(tbl_ident_q, memdb_frame(a = 4L), conflict = "ignore")
#> Matching, by = "a"
#> # Source:   [?? x 1]
#> # Database: sqlite 3.38.0 [:memory:]
#>       a
#>   <int>
#> 1     1
#> 2     2
#> 3     3
#> 4     4
rows_insert(tbl_ident_q, memdb_frame(a = 4L), conflict = "ignore", in_place = TRUE)
#> Matching, by = "a"
```

<sup>Created on 2022-04-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

## SQL Server (with #823)

``` r
library(dbplyr)
library(tidyverse)

# FIXME: Specify server connection:
con <- sql_server_con()

con_frame <- function(...) {
  copy_inline(con, tibble(...))
}
con_frame(a = 1)
#> # Source:   [?? x 1]
#> # Database: Microsoft SQL Server 15.00.4198
#>       a
#>   <dbl>
#> 1     1

DBI::dbExecute(con, "CREATE TABLE test (a INTEGER)")
DBI::dbExecute(con, "DELETE FROM test")
#> [1] 0

tbl_pure <- tbl(con, "test")
tbl_pure
#> # Source:   table<test> [?? x 1]
#> # Database: Microsoft SQL Server 15.00.4198
#> # … with 1 variable: a <int>
rows_insert(tbl_pure, con_frame(a = 1L), conflict = "ignore")
#> Matching, by = "a"
#> # Source:   [?? x 1]
#> # Database: Microsoft SQL Server 15.00.4198
#>       a
#>   <int>
#> 1     1
rows_insert(tbl_pure, con_frame(a = 1L), conflict = "ignore", in_place = TRUE)
#> Matching, by = "a"

tbl_in_schema <- tbl(con, in_schema("dbo", "test"))
tbl_in_schema
#> # Source:   table<"dbo"."test"> [?? x 1]
#> # Database: Microsoft SQL Server 15.00.4198
#>       a
#>   <int>
#> 1     1
rows_insert(tbl_in_schema, con_frame(a = 2L), conflict = "ignore")
#> Matching, by = "a"
#> # Source:   [?? x 1]
#> # Database: Microsoft SQL Server 15.00.4198
#>       a
#>   <int>
#> 1     1
#> 2     2
rows_insert(tbl_in_schema, con_frame(a = 2L), conflict = "ignore", in_place = TRUE)
#> Matching, by = "a"
#> Error in new_result(connection@ptr, statement, immediate): nanodbc/nanodbc.cpp:1594: 00000: [FreeTDS][SQL Server]The multi-part identifier ""dbo"."test".a" could not be bound.

tbl_ident <- tbl(con, ident("test"))
tbl_ident
#> # Source:   table<test> [?? x 1]
#> # Database: Microsoft SQL Server 15.00.4198
#>       a
#>   <int>
#> 1     1
rows_insert(tbl_ident, con_frame(a = 3L), conflict = "ignore")
#> Matching, by = "a"
#> # Source:   [?? x 1]
#> # Database: Microsoft SQL Server 15.00.4198
#>       a
#>   <int>
#> 1     1
#> 2     3
rows_insert(tbl_ident, con_frame(a = 3L), conflict = "ignore", in_place = TRUE)
#> Matching, by = "a"

tbl_ident_q <- tbl(con, ident_q("[test]"))
tbl_ident_q
#> # Source:   table<[test]> [?? x 1]
#> # Database: Microsoft SQL Server 15.00.4198
#>       a
#>   <int>
#> 1     1
#> 2     3
rows_insert(tbl_ident_q, con_frame(a = 4L), conflict = "ignore")
#> Matching, by = "a"
#> # Source:   [?? x 1]
#> # Database: Microsoft SQL Server 15.00.4198
#>       a
#>   <int>
#> 1     1
#> 2     3
#> 3     4
rows_insert(tbl_ident_q, con_frame(a = 4L), conflict = "ignore", in_place = TRUE)
#> Matching, by = "a"
#> Error in new_result(connection@ptr, statement, immediate): nanodbc/nanodbc.cpp:1594: 00000: [FreeTDS][SQL Server]The multi-part identifier "[test].a" could not be bound.
```

<sup>Created on 2022-04-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by krlmlr** [comment]
`sql_output_cols()` has been dropped in 71d035ddd8dfa22af9b5bc5af01cfdfe49d70217. We introduced it in dm because the `OUTPUT` clause in SQL Server is inserted in a different place from the `RETURNING` clause for other databases. It seemed like a smaller price to pay compared to reimplementing all SQL generators separately for SQL Server. What am I missing?

**Comment by krlmlr** [comment]
How do you feel about failing unconditionally with `rows_insert(conflict = "error", in_place = TRUE)`, provided that we have an escape hatch that inserts unconditionally and lets the backend decide (e.g. `conflict = "backend"` or a dedicated `rows_append()`, see https://github.com/DavisVaughan/dplyr/pull/1)?

**Comment by mgirlich** [comment]
> `sql_output_cols()` has been dropped in [71d035d](https://github.com/tidyverse/dbplyr/commit/71d035ddd8dfa22af9b5bc5af01cfdfe49d70217). We introduced it in dm because the `OUTPUT` clause in SQL Server is inserted in a different place from the `RETURNING` clause for other databases. It seemed like a smaller price to pay compared to reimplementing all SQL generators separately for SQL Server. What am I missing?
> 
> `rows_insert()` no longer works with the `returning` argument on SQL Server.

Hmm, I'm not entirely sure on this one. With `sql_output_cols()` we don't need a separate implementation for two relatively simple functions `sql_query_insert()` and `sql_query_delete()` (but we still need `sql_query_update()` and `sql_query_upsert()`). On the other hand I find it a bit confusing to have two generics for one kind of functionality (`sql_returning_cols()` and `sql_output_cols()`) and to have them both in one query.

**Comment by mgirlich** [comment]
> How do you feel about failing unconditionally with `rows_insert(conflict = "error", in_place = TRUE)`, provided that we have an escape hatch that inserts unconditionally and lets the backend decide (e.g. `conflict = "backend"` or a dedicated `rows_append()`, see [DavisVaughan/dplyr#1](https://github.com/DavisVaughan/dplyr/pull/1))?

This seems to be a good solution. Of course this implies that non-unique local tables might error quite late but the user can still do the check himself. Is there any easy to use tool for this? I know there is `dm::dm_add_pk()` but this requires to first create a dm object which seems a bit too complicated for such a simple use case.

**Comment by mgirlich** [comment]
The issue for Postgres should be fixed. I didn't expect `ident_q(set_names(cols, nms))` to be different from `set_names(ident_q(cols), nms)` but it is. Another example for issues with `ident()` and friends due to them escaping their input...

**Comment by krlmlr** [comment]
I fixed tests after #819, and added four more commits. Please feel free to revert/edit as necessary. Tests pass locally for all commits individually.

**Comment by krlmlr** [comment]
For `rows_append()` we'll need https://github.com/DavisVaughan/dplyr/pull/1 and https://github.com/tidyverse/dplyr/pull/6203, let's wait until they land in the main branch or even on CRAN.

**Comment by mgirlich** [comment]
Todo
* [ ] `rows_append()` -> wait for dplyr
* [ ] add `rows_insert_returning()` and friends and remove `returning` argument
* [ ] `rows_insert(conflict = “error”)` -> Raise an error unconditionally
* [ ] `rows_update/delete/patch(nonexistent = “error”)` -> Raise an error unconditionally
* [ ] Raise an error unconditionally – add a `check_unique` option to dbplyr only, handled by `check_dots_empty()` in dplyr?

**Comment by krlmlr** [comment]
Are we keeping the `returning` argument, or do we introduce new `rows_*_returning()` functions?

I saw that the default for `in_place` has changed to `FALSE`, I guess this means we no longer give a message if `in_place` is omitted?

Regarding integration with dm, I think we can release dbplyr first and then adapt dm -- the dbplyr update doesn't break dm's tests, because the latter still overwrites the methods. Should dbplyr give a warning mentioning that the `rows_*()` API behaves differently when dm <= 0.2.8 is loaded?

It looks like the SQL Server implementation needs more work, I'll propose something today. Postgres looks fine.

**Comment by krlmlr** [comment]
MSSQL seems good now.

**Comment by mgirlich** [comment]
> Are we keeping the `returning` argument, or do we introduce new `rows_*_returning()` functions?

I'm okay with both. Do you have a preference?

> I saw that the default for `in_place` has changed to `FALSE`, I guess this means we no longer give a message if `in_place` is omitted?

Yes, I don't see a need to give a message.
 
> Regarding integration with dm, I think we can release dbplyr first and then adapt dm -- the dbplyr update doesn't break dm's tests, because the latter still overwrites the methods. Should dbplyr give a warning mentioning that the `rows_*()` API behaves differently when dm <= 0.2.8 is loaded?

Feel free to add a message though it will probably affect only few users.

> MSSQL seems good now.

Thanks!

**Review by krlmlr — APPROVED** [review]
Thanks! Looks good to me, with https://github.com/cynkra/dm/pull/988 we get the new behavior also for the legacy code in dm. I can release dm before or shortly after dbplyr, doesn't really matter.

**Comment by krlmlr** [comment]
So, let's keep the `returning` argument and avoid adding new messages.

**Pull request merged**
